src/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -1430,8 +1430,10 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
         continue;
       }
       failureMessage = canPassThroughCanal(canalAttachment, units, player, data);
-      if (!failureMessage.isPresent()) {
```

Comment:
I don't really care how you solve the problem TBH. I do recommend the Java8 video I linked earlier. I didn't really understand 'filter' very well, the video is a nice transformation of how filtering is done in old school procedural syntax, to functional syntax that is more expressive, compact, and less error prone.